File: src/prod/src/ServiceModel/LoadMetricInformation.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace ServiceModel;

LoadMetricInformation::LoadMetricInformation()
{
}

LoadMetricInformation::LoadMetricInformation(LoadMetricInformation const & other) :
    name_(other.name_),
    isBlanacedBefore_(other.isBlanacedBefore_),
    isBlanacedAfter_(other.isBlanacedAfter_),
    deviationBefore_(other.deviationBefore_),
    deviationAfter_(other.deviationAfter_),
    balancingThreshold_(other.balancingThreshold_),
    action_(other.action_),
    activityThreshold_(other.activityThreshold_),
    clusterCapacity_(other.clusterCapacity_),
    clusterLoad_(other.clusterLoad_),
    remainingUnbufferedCapacity_(other.remainingUnbufferedCapacity_),
    nodeBufferPercentage_(other.nodeBufferPercentage_),
    bufferedCapacity_(other.bufferedCapacity_),
    remainingBufferedCapacity_(other.remainingBufferedCapacity_),
    isClusterCapacityViolation_(other.isClusterCapacityViolation_),
    minNodeLoadValue_(other.minNodeLoadValue_),
    minNodeLoadNodeId_(other.minNodeLoadNodeId_),
    maxNodeLoadValue_(other.maxNodeLoadValue_),
    maxNodeLoadNodeId_(other.maxNodeLoadNodeId_),
    currentClusterLoad_(other.currentClusterLoad_),
    bufferedClusterCapacityRemaining_(other.bufferedClusterCapacityRemaining_),
    clusterCapacityRemaining_(other.clusterCapacityRemaining_),
    maximumNodeLoad_(other.maximumNodeLoad_),
    minimumNodeLoad_(other.minimumNodeLoad_)
{
}

LoadMetricInformation::LoadMetricInformation(LoadMetricInformation && other) :
    name_(move(other.name_)),
    isBlanacedBefore_(move(other.isBlanacedBefore_)),
    isBlanacedAfter_(move(other.isBlanacedAfter_)),
    deviationBefore_(move(other.deviationBefore_)),
    deviationAfter_(move(other.deviationAfter_)),
    balancingThreshold_(move(other.balancingThreshold_)),
    action_(move(other.action_)),
    activityThreshold_(move(other.activityThreshold_)),
    clusterCapacity_(move(other.clusterCapacity_)),
    clusterLoad_(move(other.clusterLoad_)),
    remainingUnbufferedCapacity_(move(other.remainingUnbufferedCapacity_)),
    nodeBufferPercentage_(move(other.nodeBufferPercentage_)),
    bufferedCapacity_(move(other.bufferedCapacity_)),
    remainingBufferedCapacity_(move(other.remainingBufferedCapacity_)),
    isClusterCapacityViolation_(move(other.isClusterCapacityViolation_)),
    minNodeLoadValue_(move(other.minNodeLoadValue_)),
    minNodeLoadNodeId_(move(other.minNodeLoadNodeId_)),
    maxNodeLoadValue_(move(other.maxNodeLoadValue_)),
    maxNodeLoadNodeId_(move(other.maxNodeLoadNodeId_)),
    currentClusterLoad_(other.currentClusterLoad_),
    bufferedClusterCapacityRemaining_(other.bufferedClusterCapacityRemaining_),
    clusterCapacityRemaining_(other.clusterCapacityRemaining_),
    maximumNodeLoad_(other.maximumNodeLoad_),
    minimumNodeLoad_(other.minimumNodeLoad_)
{
}

LoadMetricInformation const & LoadMetricInformation::operator = (LoadMetricInformation const & other)
{
    if (this != & other)
    {
        name_ = other.name_;
        isBlanacedBefore_ = other.isBlanacedBefore_;
        isBlanacedAfter_ = other.isBlanacedAfter_;
        deviationBefore_ = other.deviationBefore_;
        deviationAfter_ = other.deviationAfter_;
        balancingThreshold_ = other.balancingThreshold_;
        action_ = other.action_;  
        activityThreshold_ = other.activityThreshold_;
        clusterCapacity_ = other.clusterCapacity_;
        clusterLoad_ = other.clusterLoad_;
        remainingUnbufferedCapacity_ = other.remainingUnbufferedCapacity_;
        nodeBufferPercentage_ = other.nodeBufferPercentage_;
        bufferedCapacity_ = other.bufferedCapacity_;
        remainingBufferedCapacity_ = other.remainingBufferedCapacity_;
        isClusterCapacityViolation_ = other.isClusterCapacityViolation_;
        minNodeLoadValue_ = other.minNodeLoadValue_;
        minNodeLoadNodeId_ = other.minNodeLoadNodeId_;
        maxNodeLoadValue_ = other.maxNodeLoadValue_;
        maxNodeLoadNodeId_ = other.maxNodeLoadNodeId_;
        currentClusterLoad_ = other.currentClusterLoad_;
        bufferedClusterCapacityRemaining_ = other.bufferedClusterCapacityRemaining_;
        clusterCapacityRemaining_ = other.clusterCapacityRemaining_;
        maximumNodeLoad_ = other.maximumNodeLoad_;
        minimumNodeLoad_ = other.minimumNodeLoad_;
    }

    return *this;
}

LoadMetricInformation const & LoadMetricInformation::operator=(LoadMetricInformation && other)
{
    if (this != &other)
    {
        name_ = move(other.name_);
        isBlanacedBefore_ = move(other.isBlanacedBefore_);
        isBlanacedAfter_ = move(other.isBlanacedAfter_);
        deviationBefore_ = move(other.deviationBefore_);
        deviationAfter_ = move(other.deviationAfter_);
        balancingThreshold_ = move(other.balancingThreshold_);
        action_ = move(other.action_);
        activityThreshold_ = move(other.activityThreshold_);
        clusterCapacity_ = move(other.clusterCapacity_);
        clusterLoad_ = move(other.clusterLoad_);
        remainingUnbufferedCapacity_ = move(other.remainingUnbufferedCapacity_);
        nodeBufferPercentage_ = move(other.nodeBufferPercentage_);
        bufferedCapacity_ = move(other.bufferedCapacity_);
        remainingBufferedCapacity_ = move(other.remainingBufferedCapacity_);
        isClusterCapacityViolation_ = move(other.isClusterCapacityViolation_);
        minNodeLoadValue_ = move(other.minNodeLoadValue_);
        minNodeLoadNodeId_ = move(other.minNodeLoadNodeId_);
        maxNodeLoadValue_ = move(other.maxNodeLoadValue_);
        maxNodeLoadNodeId_ = move(other.maxNodeLoadNodeId_);
        currentClusterLoad_ = other.currentClusterLoad_;
        bufferedClusterCapacityRemaining_ = other.bufferedClusterCapacityRemaining_;
        clusterCapacityRemaining_ = other.clusterCapacityRemaining_;
        maximumNodeLoad_ = other.maximumNodeLoad_;
        minimumNodeLoad_ = other.minimumNodeLoad_;
    }

    return *this;
}

void LoadMetricInformation::ToPublicApi(__in ScopedHeap & heap, __out FABRIC_LOAD_METRIC_INFORMATION & publicValue) const
{
    publicValue.Name = heap.AddString(name_);
    publicValue.IsBalancedAfter = isBlanacedAfter_ ? TRUE : FALSE;
    publicValue.IsBalancedBefore = isBlanacedBefore_ ? TRUE : FALSE;
    publicValue.DeviationAfter = deviationAfter_;
    publicValue.DeviationBefore = deviationBefore_;
    publicValue.BalancingThreshold = balancingThreshold_;
    publicValue.Action = heap.AddString(action_);

    auto publicValueEx1 = heap.AddItem<FABRIC_LOAD_METRIC_INFORMATION_EX1>();
    publicValueEx1->ActivityThreshold = activityThreshold_;
    publicValueEx1->ClusterCapacity = clusterCapacity_;
    publicValueEx1->ClusterLoad = clusterLoad_;
    publicValue.Reserved = publicValueEx1.GetRawPointer();

    auto publicValueEx2 = heap.AddItem<FABRIC_LOAD_METRIC_INFORMATION_EX2>();
    publicValueEx2->RemainingUnbufferedCapacity = remainingUnbufferedCapacity_;
    publicValueEx2->NodeBufferPercentage = nodeBufferPercentage_;
    publicValueEx2->BufferedCapacity = bufferedCapacity_;
    publicValueEx2->RemainingBufferedCapacity = remainingBufferedCapacity_;
    publicValueEx2->IsClusterCapacityViolation = isClusterCapacityViolation_;
    publicValueEx2->MinNodeLoadValue = minNodeLoadValue_;
    minNodeLoadNodeId_.ToPublicApi(publicValueEx2->MinNodeLoadNodeId);
    publicValueEx2->MaxNodeLoadValue = maxNodeLoadValue_;
    maxNodeLoadNodeId_.ToPublicApi(publicValueEx2->MaxNodeLoadNodeId);

    publicValueEx1->Reserved = publicValueEx2.GetRawPointer();

    auto publicValueEx3 = heap.AddItem<FABRIC_LOAD_METRIC_INFORMATION_EX3>();
    publicValueEx3->CurrentClusterLoad = currentClusterLoad_;
    publicValueEx3->BufferedClusterCapacityRemaining = bufferedClusterCapacityRemaining_;
    publicValueEx3->ClusterCapacityRemaining = clusterCapacityRemaining_;
    publicValueEx3->MaximumNodeLoad = maximumNodeLoad_;
    publicValueEx3->MinimumNodeLoad = minimumNodeLoad_;
    publicValueEx3->Reserved = NULL;

    publicValueEx2->Reserved = publicValueEx3.GetRawPointer();
}

Common::ErrorCode LoadMetricInformation::FromPublicApi(__in FABRIC_LOAD_METRIC_INFORMATION const & publicValue)
{
    auto hr = StringUtility::LpcwstrToWstring(publicValue.Name, false, name_);
    if (FAILED(hr))
    {
        return ErrorCode::FromHResult(hr);
    }

    isBlanacedAfter_ = publicValue.IsBalancedAfter ? true : false;
    isBlanacedBefore_ = publicValue.IsBalancedBefore ? true : false;
    deviationAfter_= publicValue.DeviationAfter;
    deviationBefore_ = publicValue.DeviationBefore;
    balancingThreshold_ = publicValue.BalancingThreshold;

    hr = StringUtility::LpcwstrToWstring(publicValue.Action, false, action_);
    if (FAILED(hr))
    {
        return ErrorCode::FromHResult(hr);
    }

    if (publicValue.Reserved != NULL)
    {
        FABRIC_LOAD_METRIC_INFORMATION_EX1 const & publicValueEx1 =
            *(static_cast<FABRIC_LOAD_METRIC_INFORMATION_EX1*>(publicValue.Reserved));
        activityThreshold_ = publicValueEx1.ActivityThreshold;
        clusterCapacity_ = publicValueEx1.ClusterCapacity;
        clusterLoad_ = publicValueEx1.ClusterLoad;

        if (publicValueEx1.Reserved != NULL)
        {
            FABRIC_LOAD_METRIC_INFORMATION_EX2 const & publicValueEx2 =
                *(static_cast<FABRIC_LOAD_METRIC_INFORMATION_EX2*>(publicValueEx1.Reserved));

            remainingUnbufferedCapacity_ = publicValueEx2.RemainingUnbufferedCapacity;
            nodeBufferPercentage_ = publicValueEx2.NodeBufferPercentage;
            bufferedCapacity_ = publicValueEx2.BufferedCapacity;
            remainingBufferedCapacity_ = publicValueEx2.RemainingBufferedCapacity;
            isClusterCapacityViolation_ = publicValueEx2.IsClusterCapacityViolation ? true : false;
            minNodeLoadValue_ = publicValueEx2.MinNodeLoadValue;
            minNodeLoadNodeId_.FromPublicApi(publicValueEx2.MinNodeLoadNodeId);
            maxNodeLoadValue_ = publicValueEx2.MaxNodeLoadValue;
            maxNodeLoadNodeId_.FromPublicApi(publicValueEx2.MaxNodeLoadNodeId);

            if (publicValueEx2.Reserved != NULL)
            {
                FABRIC_LOAD_METRIC_INFORMATION_EX3 const & publicValueEx3 =
                    *(static_cast<FABRIC_LOAD_METRIC_INFORMATION_EX3*>(publicValueEx2.Reserved));

                currentClusterLoad_ = publicValueEx3.CurrentClusterLoad;
                bufferedClusterCapacityRemaining_ = publicValueEx3.BufferedClusterCapacityRemaining;
                clusterCapacityRemaining_ = publicValueEx3.ClusterCapacityRemaining;
                maximumNodeLoad_ = publicValueEx3.MaximumNodeLoad;
                minimumNodeLoad_ = publicValueEx3.MinimumNodeLoad;
            }
        }
    }

    return ErrorCode(ErrorCodeValue::Success);
}

void LoadMetricInformation::WriteTo(Common::TextWriter& w, Common::FormatOptions const &) const
{
    w.Write(ToString());
}

std::wstring LoadMetricInformation::ToString() const
{
    wstring objectString;
    auto error = JsonHelper::Serialize(const_cast<LoadMetricInformation&>(*this), objectString);
    if (!error.IsSuccess())
    {
        return wstring();
    }
    return objectString;
}

File: src/prod/src/managed/Api/src/System/Fabric/query/LoadMetricInformation.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Query
{
    using System.Collections.Generic;
    using System.Fabric.Common.Serialization;
    using System.Fabric.Interop;
    using System.Text;

    /// <summary>
    /// <para>Represents the load metric information.</para>
    /// </summary>
    public sealed class LoadMetricInformation
    {
        /// <summary>
        /// <para>Initializes a new instance of <see cref="System.Fabric.Query.LoadMetricInformation"/>.</para>
        /// </summary>
        public LoadMetricInformation()
        {
        }

        /// <summary>
        /// <para>Gets the name of the metric.</para>
        /// </summary>
        /// <value>
        /// <para>The name of the metric.</para>
        /// </value>
        public string Name { get; internal set; }

        /// <summary>
        /// <para>Gets a value that indicates whether the metrics is balanced or not before resource balancer run.</para>
        /// </summary>
        /// <value>
        /// <para>
        ///     <languageKeyword>true</languageKeyword> if the metrics is balanced or not before resource balancer run; 
        /// otherwise, <languageKeyword>false</languageKeyword>.</para>
        /// </value>
        public bool IsBalancedBefore { get; internal set; }

        /// <summary>
        /// <para>Gets a value that indicates whether the metrics is balanced or not after resource balancer run.</para>
        /// </summary>
        /// <value>
        /// <para>
        ///     <languageKeyword>true</languageKeyword> if the metrics is balanced or not after resource balancer run; 
        /// otherwise, <languageKeyword>false</languageKeyword>.</para>
        /// </value>
        public bool IsBalancedAfter { get; internal set; }

        /// <summary>
        /// <para>Gets the standard average deviation of the metrics before resource balancer run.</para>
        /// </summary>
        /// <value>
        /// <para>The standard average deviation of the metrics before resource balancer run.</para>
        /// </value>
        public double DeviationBefore { get; internal set; }

        /// <summary>
        /// <para>Gets the standard average deviation of the metrics after resource balancer run.</para>
        /// </summary>
        /// <value>
        /// <para>The standard average deviation of the metrics after resource balancer run.</para>
        /// </value>
        public double DeviationAfter { get; internal set; }

        /// <summary>
        /// <para>Gets he balancing threshold for a certain metric.</para>
        /// </summary>
        /// <value>
        /// <para>The balancing threshold for a certain metric.</para>
        /// </value>
        public double BalancingThreshold { get; internal set; }

        /// <summary>
        /// <para>Gets the current action being taken with regard to this metric. Examples could be placement, balancing, and constraint checks.</para>
        /// </summary>
        /// <value>
        /// <para>The current action being taken with regard to this metric.</para>
        /// </value>
        public string Action { get; internal set; }

        /// <summary>
        /// <para>Gets the Activity Threshold specified for this metric in the system Cluster Manifest.</para>
        /// </summary>
        /// <value>
        /// <para>The Activity Threshold specified for this metric in the system Cluster Manifest.</para>
        /// </value>
        public long ActivityThreshold { get; internal set; }

        /// <summary>
        /// <para>Gets the total cluster capacity for a given metric, determined by summing the capacity for the metric across all nodes.</para>
        /// </summary>
        /// <value>
        /// <para>The total cluster capacity for a given metric.</para>
        /// </value>
        public long ClusterCapacity { get; internal set; }

        /// <summary>
        /// <para>Gets the total cluster load for a specific metric within the cluster</para>
        /// <para>
        /// In future releases of Service Fabric this parameter will be deprecated in favor of
        /// <see cref="System.Fabric.Query.LoadMetricInformation.CurrentClusterLoad"/>.
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The total cluster load.</para>
        /// </value>
        public long ClusterLoad { get; internal set; }

        /// <summary>
        /// <para>Gets the total cluster load for a specific metric within the cluster.</para>
        /// </summary>
        /// <value>
        /// <para>The total cluster load.</para>
        /// </value>
        public double CurrentClusterLoad { get; internal set; }

        /// <summary>
        /// <para>Gets remaining capacity for the metric in the cluster. Remaining capacity is defined as the current Cluster Capacity 
        /// minus the current Cluster Load.</para>
        /// <para>
        /// In future releases of Service Fabric this parameter will be deprecated in favor of
        /// <see cref="System.Fabric.Query.LoadMetricInformation.ClusterCapacityRemaining"/>.
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The remaining capacity for the metric in the cluster.</para>
        /// </value>
        [JsonCustomization(PropertyName = JsonPropertyNames.RemainingUnbufferedCapacity)]
        public long ClusterRemainingCapacity { get; internal set; }

        /// <summary>
        /// <para>Gets remaining capacity for the metric in the cluster. Remaining capacity is defined as the current Cluster Capacity 
        /// minus the current Cluster Load.</para>
        /// </summary>
        /// <value>
        /// <para>The remaining capacity for the metric in the cluster.</para>
        /// </value>
        public double ClusterCapacityRemaining { get; internal set; }

        /// <summary>
        /// <para>Gets whether the metric is currently over capacity in the cluster.</para>
        /// </summary>
        /// <value>
        /// <para>
        ///     <languageKeyword>true</languageKeyword> if the metric is currently over capacity in the cluster; 
        /// otherwise, <languageKeyword>false</languageKeyword>.</para>
        /// </value>
        public bool IsClusterCapacityViolation { get; internal set; }

        /// <summary>
        /// <para>
        /// Gets the reserved percentage of total node capacity for this metric
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The reserved percentage of total node capacity for this metric</para>
        /// </value>
        public double NodeBufferPercentage { get; internal set; }

        /// <summary>
        /// <para>
        /// The reserved percentage of cluster total capacity for this metric
        /// </para>
        /// <para>
        /// In future releases of Service Fabric this parameter will be deprecated in favor of
        /// <see cref="System.Fabric.Query.LoadMetricInformation.BufferedClusterCapacityRemaining"/>.
        /// </para>
        /// </summary>
        /// <value>
        /// <para>ClusterBufferedCapacity property gets/sets the value of the long field, _clusterBufferedCapacity.</para>
        /// </value>
        [JsonCustomization(PropertyName = JsonPropertyNames.BufferedCapacity)]
        public long ClusterBufferedCapacity { get; internal set; }

        /// <summary>
        /// <para>
        /// The reserved percentage of cluster total capacity for this metric
        /// </para>
        /// </summary>
        /// <value>
        /// <para>Remaining capacity in the cluster excluding the reserved space.</para>
        /// </value>
        public double BufferedClusterCapacityRemaining { get; internal set; }

        /// <summary>
        /// <para>
        /// Gets the remaining percentage of cluster total capacity for this metric
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The remaining percentage of cluster total capacity for this metric.</para>
        /// </value>
        [JsonCustomization(PropertyName = JsonPropertyNames.RemainingBufferedCapacity)]
        public long ClusterRemainingBufferedCapacity { get; internal set; }

        /// <summary>
        /// <para>
        /// Gets the minimum load on any node for this metric.
        /// </para>
        /// <para>
        /// In future releases of Service Fabric this parameter will be deprecated in favor of
        /// <see cref="System.Fabric.Query.LoadMetricInformation.MinimumNodeLoad"/>.
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The minimum load on any node for this metric.</para>
        /// </value>
        public long MinNodeLoadValue { get; internal set; }

        /// <summary>
        /// <para>
        /// Gets the minimum load on any node for this metric.
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The minimum load on any node for this metric.</para>
        /// </value>
        public double MinimumNodeLoad { get; internal set; }

        /// <summary>
        /// <para>
        /// Gets the node id of the node with the minimum load for this metric
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The node id of the node with the minimum load for this metric.</para>
        /// </value>
        [JsonCustomization(PropertyName = JsonPropertyNames.MinNodeLoadId)]
        public NodeId MinNodeLoadNodeId { get; internal set; }

        /// <summary>
        /// <para>
        /// Gets the maximum load on any node for this metric.
        /// </para>
        /// <para>
        /// In future releases of Service Fabric this parameter will be deprecated in favor of
        /// <see cref="System.Fabric.Query.LoadMetricInformation.MaximumNodeLoad"/>.
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The maximum load on any node for this metric.</para>
        /// </value>
        public long MaxNodeLoadValue { get; internal set; }

        /// <summary>
        /// <para>
        /// Gets the maximum load on any node for this metric.
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The maximum load on any node for this metric.</para>
        /// </value>
        public double MaximumNodeLoad { get; internal set; }

        /// <summary>
        /// <para>
        /// Gets the node id of the node with the maximum load for this metric
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The node id of the node with the maximum load for this metric.</para>
        /// </value>
        [JsonCustomization(PropertyName = JsonPropertyNames.MaxNodeLoadId)]
        public NodeId MaxNodeLoadNodeId { get; internal set; }

        internal static unsafe LoadMetricInformation CreateFromNative(NativeTypes.FABRIC_LOAD_METRIC_INFORMATION nativeResultItem)
        {
            long activityThreshold = 0;
            long clusterCapacity = 0;
            long clusterLoad = 0;
            long clusterRemainingCapacity = 0;
            double nodeBufferPercentage = 0;
            long clusterBufferedCapacity = 0;
            long clusterRemainingBufferedCapacity = 0;
            bool isClusterCapacityViolation = false;
            long minNodeLoadValue = -1;
            NodeId minNodeLoadNodeId = null;
            long maxNodeLoadValue = -1;
            NodeId maxNodeLoadNodeId = null;
            double currentClusterLoad = 0;
            double bufferedClusterCapacityRemaining = 0;
            double clusterCapacityRemaining = 0;
            double maximumNodeLoad = -1;
            double minimumNodeLoad = -1;

            if (nativeResultItem.Reserved != IntPtr.Zero)
            {
                NativeTypes.FABRIC_LOAD_METRIC_INFORMATION_EX1* ext1 = (NativeTypes.FABRIC_LOAD_METRIC_INFORMATION_EX1*)nativeResultItem.Reserved;
                activityThreshold = ext1->ActivityThreshold;
                clusterCapacity = ext1->ClusterCapacity;
                clusterLoad = ext1->ClusterLoad;

                // calculate remaining capacity and isCapacityViolation flag
                if (clusterCapacity != 0)
                {
                    if (clusterCapacity < clusterLoad)
                    {
                        isClusterCapacityViolation = true;
                        clusterRemainingCapacity = 0;
                    }
                    else
                    {
                        clusterRemainingCapacity = clusterCapacity - clusterLoad;
                    }
                }

                if (ext1->Reserved != IntPtr.Zero)
                {
                    NativeTypes.FABRIC_LOAD_METRIC_INFORMATION_EX2* ext2 = (NativeTypes.FABRIC_LOAD_METRIC_INFORMATION_EX2*)ext1->Reserved;

                    clusterRemainingCapacity = ext2->RemainingUnbufferedCapacity;
                    nodeBufferPercentage = ext2->NodeBufferPercentage;
                    clusterBufferedCapacity = ext2->BufferedCapacity;
                    clusterRemainingBufferedCapacity = ext2->RemainingBufferedCapacity;
                    isClusterCapacityViolation = NativeTypes.FromBOOLEAN(ext2->IsClusterCapacityViolation);
                    minNodeLoadValue = ext2->MinNodeLoadValue;
                    minNodeLoadNodeId = NativeTypes.FromNativeNodeId(ext2->MinNodeLoadNodeId);
                    maxNodeLoadValue = ext2->MaxNodeLoadValue;
                    maxNodeLoadNodeId = NativeTypes.FromNativeNodeId(ext2->MaxNodeLoadNodeId);

                    if (ext2->Reserved != IntPtr.Zero)
                    {
                        NativeTypes.FABRIC_LOAD_METRIC_INFORMATION_EX3* ext3 = (NativeTypes.FABRIC_LOAD_METRIC_INFORMATION_EX3*)ext2->Reserved;
                        currentClusterLoad = ext3->CurrentClusterLoad;
                        bufferedClusterCapacityRemaining = ext3->BufferedClusterCapacityRemaining;
                        clusterCapacityRemaining = ext3->ClusterCapacityRemaining;
                        maximumNodeLoad = ext3->MaximumNodeLoad;
                        minimumNodeLoad = ext3->MinimumNodeLoad;
                    }
                }
            }

            return new LoadMetricInformation
            {
                Name = NativeTypes.FromNativeString(nativeResultItem.Name),
                IsBalancedBefore = NativeTypes.FromBOOLEAN(nativeResultItem.IsBalancedBefore),
                IsBalancedAfter = NativeTypes.FromBOOLEAN(nativeResultItem.IsBalancedAfter),
                DeviationBefore = nativeResultItem.DeviationBefore,
                DeviationAfter = nativeResultItem.DeviationAfter,
                BalancingThreshold = nativeResultItem.BalancingThreshold,
                Action = NativeTypes.FromNativeString(nativeResultItem.Action),
                ActivityThreshold = activityThreshold,
                ClusterCapacity = clusterCapacity,
                ClusterLoad = clusterLoad,
                ClusterRemainingCapacity = clusterRemainingCapacity,
                NodeBufferPercentage = nodeBufferPercentage,
                ClusterBufferedCapacity = clusterBufferedCapacity,
                ClusterRemainingBufferedCapacity = clusterRemainingBufferedCapacity,
                IsClusterCapacityViolation = isClusterCapacityViolation,
                MinNodeLoadValue = minNodeLoadValue,
                MinNodeLoadNodeId = minNodeLoadNodeId,
                MaxNodeLoadValue = maxNodeLoadValue,
                MaxNodeLoadNodeId = maxNodeLoadNodeId,
                CurrentClusterLoad = currentClusterLoad,
                BufferedClusterCapacityRemaining = bufferedClusterCapacityRemaining,
                ClusterCapacityRemaining = clusterCapacityRemaining,
                MaximumNodeLoad = maximumNodeLoad,
                MinimumNodeLoad = minimumNodeLoad
        };
        }

        internal static unsafe IList<LoadMetricInformation> CreateFromNativeList(NativeTypes.FABRIC_LOAD_METRIC_INFORMATION_LIST* nativeResultList)
        {
            var rv = new List<LoadMetricInformation>();

            var nativeArray = (NativeTypes.FABRIC_LOAD_METRIC_INFORMATION*)nativeResultList->Items;
            for (int i = 0; i < nativeResultList->Count; i++)
            {
                var nativeItem = *(nativeArray + i);
                rv.Add(LoadMetricInformation.CreateFromNative(nativeItem));
            }

            return rv;
        }

        /// <summary>
        /// <para>
        /// Pretty print out each field of <see cref="System.Fabric.Query.LoadMetricInformation" />.
        /// </para>
        /// </summary>
        /// <returns>
        /// <para>A string representing the load metric information.</para>
        /// </returns>
        /// <example>
        ///     LoadMetricName        : Metric1
        ///     IsBalancedBefore      : True
        ///     IsBalancedAfter       : True
        ///     DeviationBefore       : 2
        ///     DeviationAfter        : 2
        ///     BalancingThreshold    : 1
        ///     Action    : NoActionNeeded
        ///     ActivityThreshold     : 3
        ///     ClusterCapacity       : 100
        ///     CurrentClusterLoad    : 1.0
        ///     ClusterCapacityRemaining : 0.0
        ///     NodeBufferPercentage  : 0
        ///     ClusterBufferedCapacity : 0
        ///     BufferedClusterCapacityRemaining : 0.0
        ///     ClusterCapacityViolation : True
        ///     MaximumNodeLoad       : 1.0
        ///     MinNodeLoadNodeId     : 1ca9521d70301383417536df0c96f671
        ///     MinimumNodeLoad       : 0.0
        ///     MaxNodeLoadNodeId     : cf68563e16a44f808e86197a9cf83de5
        /// </example>
        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();

            sb.AppendFormat("Name: {0}\n", Name);
            sb.AppendFormat("IsBalancedBefore: {0}\n", IsBalancedBefore);
            sb.AppendFormat("IsBalancedAfter: {0}\n", IsBalancedAfter);
            sb.AppendFormat("DeviationBefore: {0}\n", DeviationBefore);
            sb.AppendFormat("DeviationAfter: {0}\n", DeviationAfter);
            sb.AppendFormat("BalancingThreshold: {0}\n", BalancingThreshold);
            sb.AppendFormat("Action: {0}\n", Action);
            sb.AppendFormat("ActivityThreshold: {0}\n", ActivityThreshold);
            sb.AppendFormat("ClusterCapacity: {0}\n", ClusterCapacity);
            sb.AppendFormat("CurrentClusterLoad: {0}\n", CurrentClusterLoad);
            sb.AppendFormat("ClusterCapacityRemaining: {0}\n", ClusterCapacityRemaining);
            sb.AppendFormat("NodeBufferPercentage: {0}\n", NodeBufferPercentage);
            sb.AppendFormat("ClusterBufferedCapacity: {0}\n", ClusterBufferedCapacity);
            sb.AppendFormat("BufferedClusterCapacityRemaining: {0}\n", BufferedClusterCapacityRemaining);
            sb.AppendFormat("IsClusterCapacityViolation: {0}\n", IsClusterCapacityViolation);
            sb.AppendFormat("MinimumNodeLoad: {0}\n", MinimumNodeLoad);
            sb.AppendFormat("MinNodeLoadNodeId: {0}\n", MinNodeLoadNodeId);
            sb.AppendFormat("MaximumNodeLoad: {0}\n", MaximumNodeLoad);
            sb.AppendFormat("MaxNodeLoadNodeId: {0}\n", MaxNodeLoadNodeId);

            return sb.ToString();
        }
    }
}